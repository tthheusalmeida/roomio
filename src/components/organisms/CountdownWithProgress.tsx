"use client";

import { useState } from "react";
import ProgressBar from "@/components/atoms/ProgressBar";
import CountdownTimer from "@/components/organisms/CountdownTimer";

interface CountdownWithProgressProps {
  initialMinutes: number;
  onFinish?: () => void;
  className?: string;
  height?: string;
  color?: string;
  showLabel?: boolean;
  mode?: "both" | "bar" | "timer";
  shouldStop?: boolean;
}

export default function CountdownWithProgress({
  initialMinutes,
  onFinish,
  className,
  height = "h-2",
  color = "bg-green-600",
  showLabel = true,
  mode = "both",
  shouldStop,
}: CountdownWithProgressProps) {
  const [progress, setProgress] = useState(0);

  return (
    <div className={`w-full space-y-1 ${className}`}>
      {(mode === "bar" || mode === "both") && (
        <ProgressBar
          value={progress}
          height={height}
          color={color}
          showLabel={showLabel}
        />
      )}

      {(mode === "timer" || mode === "both") && (
        <CountdownTimer
          initialMinutes={initialMinutes}
          onFinish={onFinish}
          onProgress={setProgress}
          className="text-center"
          shouldStop={shouldStop}
        />
      )}
    </div>
  );
}
