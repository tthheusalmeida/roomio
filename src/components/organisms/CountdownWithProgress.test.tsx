import { render, screen, act } from "@testing-library/react";
import CountdownWithProgress from "./CountdownWithProgress";

jest.useFakeTimers();

describe("CountdownWithProgress", () => {
  afterEach(() => {
    jest.clearAllTimers();
  });

  it("renders both ProgressBar and CountdownTimer by default", () => {
    render(<CountdownWithProgress initialMinutes={0.05} />);

    const progressBar = screen.getByText("0%").closest("div");
    expect(progressBar).toBeInTheDocument();

    expect(screen.getByText(/00:0[1-3]/)).toBeInTheDocument();
  });

  it("renders only ProgressBar when mode is 'bar'", () => {
    render(<CountdownWithProgress initialMinutes={1} mode="bar" />);

    const progressBar = screen.getByText("0%").closest("div");
    expect(progressBar).toBeInTheDocument();

    expect(screen.queryByText(/:/)).not.toBeInTheDocument();
  });

  it("renders only CountdownTimer when mode is 'timer'", () => {
    render(<CountdownWithProgress initialMinutes={1} mode="timer" />);

    expect(screen.queryByText(/%/)).not.toBeInTheDocument();

    expect(screen.getByText("01:00")).toBeInTheDocument();
  });

  it("updates progress and calls onFinish", () => {
    const onFinish = jest.fn();
    render(<CountdownWithProgress initialMinutes={0.05} onFinish={onFinish} />);

    // AvanÃ§a tempo
    act(() => {
      jest.advanceTimersByTime(1500);
    });

    act(() => {
      jest.advanceTimersByTime(3000);
    });

    expect(onFinish).toHaveBeenCalled();
  });

  it("matches snapshot", () => {
    const { container } = render(<CountdownWithProgress initialMinutes={1} />);
    expect(container).toMatchSnapshot();
  });
});
