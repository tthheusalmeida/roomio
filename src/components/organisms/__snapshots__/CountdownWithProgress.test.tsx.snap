// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`CountdownWithProgress matches snapshot 1`] = `
<div>
  <div
    class="w-full space-y-1 undefined"
  >
    <div
      class="w-full bg-gray-200 rounded h-2 overflow-hidden"
    >
      <div
        class="transition-all duration-300 bg-green-600 h-2"
        style="width: 0%;"
      />
      <p
        class="text-sm text-center text-gray-500 mt-1"
      >
        0
        %
      </p>
    </div>
    <div
      class="text-center"
    >
      <p>
        01:00
      </p>
    </div>
  </div>
</div>
`;
